### PR TITLE
Integrate snapshot export step in CI

### LINKED ISSUE #2195: Test snapshot exporting during CI
**Issue summary**

We have a CI step that imports a calibnet snapshot and verifies that it can correctly sync to the most recent tipset. We should extend this stop by also exporting a snapshot. Even as calibnet grows larger, this should take no more than a few seconds.

Snapshot exporting is also done one per day on an external server but we want quicker feedback when something stops working.

**Other information and links**

### PR DESCRIPTION
**Pull request by creativcoder**
**Summary of changes**
Changes introduced in this pull request:
- [X] Refactors walk_snapshot method
- [X] Adds snapshot export step in CI

**Reference issue to close (if applicable)**

Closes #2195 

**Other information and links**

### CODE REVIEW

**Comment by lemmih** [comment]
I know it's still a draft but I'm not in love with the organization. Moving all the `ipld` utils into `chain_store` feels awkward.

Also, you're mixing async code and sync code in an unusual way.

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -137,6 +137,17 @@ jobs:
         run: |
           forest-cli sync wait
           forest-cli --chain calibnet db stats
+      - name: export snapshot and sync forest with exported snapshot
+        run: |
+          rm /tmp/snapshots/*
+          forest-cli snapshot export -o /tmp/snapshots
+          forest --chain calibnet --target-peer-count 50 --encrypt-keystore false --halt-after-import --import-snapshot /tmp/snapshots/*.car
```
Comment (creativcoder) [inline .github/workflows/rust.yml:144]: I wonder if we can reuse steps within a job, by referring to their ids. Is this possible?

**Comment by creativcoder** [comment]
> I know it's still a draft but I'm not in love with the organization. Moving all the `ipld` utils into `chain_store` feels awkward.
> 
> Also, you're mixing async code and sync code in an unusual way.

This has been refactored as discussed

File: .github/workflows/rust.yml (hunk 1 above)
Comment (lemmih) [reply .github/workflows/rust.yml:144]: You can definitely re-use the database that was created in the previous steps. No need to import another snapshot.

**Review by lemmih — CHANGES_REQUESTED** [review]
I'm not completely convinced about the `SnapshotWalker` structure. Why not use a callback like in the old `walk_snapshot`?

**Comment by creativcoder** [comment]
> I'm not completely convinced about the `SnapshotWalker` structure. Why not use a callback like in the old `walk_snapshot`?

I tried refactoring back to closure based fn, but i got a few closure capture related errors:
https://github.com/ChainSafe/forest/commit/38d3d5e5f0a5e743fbf0b6cbe77b43c85774b847  thoughts?

Besides that, the main reasons I went for this refactor were:
1. Closure was doing two things, getting the blockstore and then sending the block. It just felt it was doing a lot.
2. Readability and contributer friendly. It can be hard to follow callback based code sometimes.

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -144,6 +145,11 @@ jobs:
           echo "--- Forest STDOUT ---"; cat forest.out
           echo "--- Forest STDERR ---"; cat forest.err
           echo "--- Forest Prometheus metrics ---"; cat metrics.log
+      - name: export snapshot and sync forest with exported snapshot
+        run: |
+          rm $SNAPSHOT_DIRECTORY/*
+          forest-cli snapshot export -o $SNAPSHOT_DIRECTORY
```
Comment (lemmih) [inline .github/workflows/rust.yml:151]: You can't export a snapshot without having a running node.

Comment (lemmih) [reply .github/workflows/rust.yml:151]: `forest-cli` does pretty much nothing on its own. It just sends messages to the Forest node.

**Comment by lemmih** [comment]
> > I'm not completely convinced about the `SnapshotWalker` structure. Why not use a callback like in the old `walk_snapshot`?
> 
> I tried refactoring back to closure based fn, but i got a few closure capture related errors: [38d3d5e](https://github.com/ChainSafe/forest/commit/38d3d5e5f0a5e743fbf0b6cbe77b43c85774b847) thoughts?
> 
> Besides that, the main reasons I went for this refactor were:
> 
>     1. Closure was doing two things, getting the blockstore and then sending the block. It just felt it was doing a lot.
> 
>     2. Readability and contributer friendly. It can be hard to follow callback based code sometimes.

I'm not sure the control-flow of `SnapshotWalker` is easier to follow than that of `walk_snapshot`. And, for better or worse, we're already using the callback pattern in `recurse_links`.

File: blockchain/chain/src/store/chain_store.rs [hunk 3]
```diff
@@ -617,18 +594,63 @@ where
...
+    }
+
+    async fn send_block(&self, cid: Cid) -> Result<(), Error> {
+        let block = self.load(cid)?;
+        self.tx
+            .send_async((cid, block))
+            .await
+            .expect("failed sending block for export");
+        Ok(())
+    }
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:642]: Avoiding cloning a block is not worth having duplicated code.

File: blockchain/chain/src/store/chain_store.rs [hunk 4]
```diff
@@ -617,18 +594,63 @@ where
...
+
+    async fn load_and_send(&self, cid: Cid) -> Result<Vec<u8>, anyhow::Error> {
+        let block = self.load(cid)?;
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:645]: `self.load` can be inlined here.

File: blockchain/chain/src/store/chain_store.rs [hunk 5]
```diff
@@ -617,18 +594,63 @@ where
...
+        self.tx
+            .send_async((cid, block.clone()))
+            .await
+            .expect("failed sending block for export");
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:649]: Can we use `?` here?

File: blockchain/chain/src/store/chain_store.rs [hunk 6]
```diff
@@ -650,8 +672,10 @@ where
                 }
             }
 
-            if !skip_old_msgs || h.epoch() > incl_roots_epoch {
-                recurse_links(&mut seen, *h.messages(), &mut load_block)?;
+            let mut loader = |cid| Box::pin(self.load_and_send(cid));
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:675]:
```suggestion
            let mut loader = |cid| self.load_and_send(cid);
```

**Review by lemmih — CHANGES_REQUESTED** [review]

**Comment by creativcoder** [comment]
> I'm not sure the control-flow of `SnapshotWalker` is easier to follow than that of `walk_snapshot`. And, for better or worse, we're already using the callback pattern in `recurse_links`.

Reverted back to closure based walker fn

File: blockchain/chain/src/store/chain_store.rs (hunk 6 above)
Comment (creativcoder) [reply blockchain/chain/src/store/chain_store.rs]: Fixed.

File: blockchain/chain/src/store/chain_store.rs (hunk 5 above)
Comment (creativcoder) [reply blockchain/chain/src/store/chain_store.rs]: Fixed.

File: blockchain/chain/src/store/chain_store.rs (hunk 4 above)
Comment (creativcoder) [reply blockchain/chain/src/store/chain_store.rs]: Fixed.

File: blockchain/chain/src/store/chain_store.rs (hunk 3 above)
Comment (creativcoder) [reply blockchain/chain/src/store/chain_store.rs]: Fixed.

File: .github/workflows/rust.yml (hunk 2 above)
Comment (creativcoder) [reply .github/workflows/rust.yml]: ah i reordered this step incorrectly. fixed.

File: blockchain/chain/src/store/chain_store.rs [hunk 7]
```diff
@@ -492,30 +493,24 @@ where
...
-            });
-
-            Ok(block)
-        })
-        .await?;
+            async move {
+                let block = self.blockstore().get(&cid)?.ok_or_else(|| {
+                    if skip_old_msgs {
+                        Error::Other("Cid {cid} not found in blockstore".to_string())
+                    } else {
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:501]: I think we should get rid of `skip_old_msgs`. Lotus is also moving away from "fat" snapshots ("fat" snapshots were used by block-explorers in the past but they can download the message receipts in other ways).

Not sure if removing the flag is appropriate for this PR or if it should be in a separate PR.

File: blockchain/chain/src/store/chain_store.rs [hunk 8]
```diff
@@ -534,6 +529,61 @@ where
         Ok(digest)
     }
 
+    /// Walks over tipset and state data and loads all blocks not yet seen.
+    /// This is tracked based on the callback function loading blocks.
+    async fn walk_snapshot<F, T>(
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:534]: If you move this function down a few lines then the diff will be prettier.

File: blockchain/chain/src/store/chain_store.rs (hunk 7 above)
Comment (creativcoder) [reply blockchain/chain/src/store/chain_store.rs:501]: I see, can remove it but removing `skip_old_msgs` would touch other modules and probably make this PR big. I can do that in a follow up PR.

File: blockchain/chain/src/store/chain_store.rs (hunk 8 above)
Comment (creativcoder) [reply blockchain/chain/src/store/chain_store.rs:534]: Done

**Review by lemmih — APPROVED** [review]

File: .github/workflows/rust.yml [hunk 9]
```diff
@@ -135,17 +136,23 @@ jobs:
...
-          forest --chain calibnet --target-peer-count 50 --encrypt-keystore false --halt-after-import --import-snapshot /tmp/snapshots/*.car
+          forest --chain calibnet --target-peer-count 50 --encrypt-keystore false --halt-after-import --import-snapshot $SNAPSHOT_DIRECTORY/*.car
           forest-cli --chain calibnet db stats
           forest --chain calibnet --target-peer-count 50 --encrypt-keystore false --detach
       - name: wait for sync and check health
         run: |
           forest-cli sync wait
           forest-cli --chain calibnet db stats
-      - name: kill Forest process, print forest logs
+      - name: export snapshot and sync forest with exported snapshot
```
Comment (LesnyRumcajs) [inline .github/workflows/rust.yml:149]: It doesn't sync in this step, no? It imports the snapshot. If we want to sync, we need to `sync wait` like above.

Also, it may be a good exercise to check if the `shasum` of the exported snapshot is correct.

**Review by LesnyRumcajs — APPROVED** [review]
Looks good in general. The CI steps are getting kind of huge though, it would be fantastic if we could move them to a separate Makefile and only call `make ci-blabla` from the workflow. It would make the workflow files way more readable and CI steps easier to reproduce locally.

File: .github/workflows/rust.yml (hunk 9 above)
Comment (creativcoder) [reply .github/workflows/rust.yml]: We decided with @lemmih that just snapshot export would do. We would later add a snapshot verify command to validate exported snapshots.

Comment (creativcoder) [reply .github/workflows/rust.yml]: I've added the sha256 check step in CI
